LibreNMS/OS/Vrp.php

```diff
@@ -404,28 +404,52 @@ public function discoverWirelessApCount()
     public function discoverWirelessClients()
     {
         $sensors = [];
-        $total_oids = [];
-
-        $vapInfoTable = $this->getCacheTable('hwWlanVapInfoTable', 'HUAWEI-WLAN-VAP-MIB', 3);
-
-        foreach ($vapInfoTable as $a_index => $ap) {
-            //Convert mac address (hh:hh:hh:hh:hh:hh) to dec OID (ddd.ddd.ddd.ddd.ddd.ddd)
-            $a_index_oid = implode('.', array_map('hexdec', explode(':', $a_index)));
-            foreach ($ap as $r_index => $radio) {
-                foreach ($radio as $s_index => $ssid) {
-                    $oid = '.1.3.6.1.4.1.2011.6.139.17.1.1.1.9.' . $a_index_oid . '.' . $r_index . '.' . $s_index;
-                    $total_oids[] = $oid;
-                    $sensors[] = new WirelessSensor(
-                        'clients',
-                        $this->getDeviceId(),
-                        $oid,
-                        'vrp',
-                        $a_index_oid . '.' . $r_index . '.' . $s_index,
-                        'Radio:' . $r_index . ' SSID:' . $ssid['hwWlanVapProfileName'],
-                        $ssid['hwWlanVapStaOnlineCnt']
-                    );
-                }
+
+        //Clients (STA) for 2.4Ghz and 5Ghz
+        $sta5gTable = $this->getCacheTable('hwWlanSsid5gStaCnt', 'HUAWEI-WLAN-VAP-MIB', 3);
+        $sta2gTable = $this->getCacheTable('hwWlanSsid2gStaCnt', 'HUAWEI-WLAN-VAP-MIB', 3);
+        $staTable = array_merge_recursive($sta5gTable, $sta2gTable);
```

Comment:
snmpwalk_group() can accept an input array so you don't have to merge them.  Also, it doesn't look like this data is used for any other sensor, so no need to cache them.

Reply:
Makes sense. Will fix that.

Reply:
Hum, snmpwalk_group only works with one OID, not an array. So the merge is still mandatory. But at least caching will be avoided.

Reply:
nope, like this (changed to snmpwalk_cache_oid because group does not allow specifying snmp flags):
```php
        $staTable = snmpwalk_cache_oid($this->getDeviceArray(), 'hwWlanSsid', [], 'HUAWEI-WLAN-VAP-MIB', null, '-OQUsb');
        $staTable = snmpwalk_cache_oid($this->getDeviceArray(), 'hwWlanSsid2gStaCnt', $staTable, 'HUAWEI-WLAN-VAP-MIB', null, '-OQUsb');
        $staTable = snmpwalk_cache_oid($this->getDeviceArray(), 'hwWlanSsid2gStaCnt', $staTable, 'HUAWEI-WLAN-VAP-MIB', null, '-OQUsb');

```

Reply:
See how passing the same array inserts them into it.  Much better than a recursive merge.

Reply:
`        $staTable = snmpwalk_cache_oid($this->getDeviceArray(), 'hwWlanSsid', [], 'HUAWEI-WLAN-VAP-MIB', null, '-OQUsb');`
But `hwWlanSsid` is the index. It can't have it in the result of any snmpwalk. So this line returns an empty array. Only the hwWlanSsid2gStaCnt and hwWlanSsid5gStaCnt give a result. So no matter what, I'll need to do the translation to get num_to_string or string_to_num. 

So at the end, here are the 2 working variants:
```
        $sta5gTable = snmpwalk_group($this->getDeviceArray(), 'hwWlanSsid5gStaCnt', 'HUAWEI-WLAN-VAP-MIB', 3);
        $sta2gTable = snmpwalk_group($this->getDeviceArray(), 'hwWlanSsid2gStaCnt', 'HUAWEI-WLAN-VAP-MIB', 3);
        $staTable = array_merge_recursive($sta5gTable, $sta2gTable);
```
and 
```
        $staTable = snmpwalk_cache_oid($this->getDeviceArray(), 'hwWlanSsid2gStaCnt', [], 'HUAWEI-WLAN-VAP-MIB', null, '-OQUs');
        $staTable = snmpwalk_cache_oid($this->getDeviceArray(), 'hwWlanSsid5gStaCnt', $staTable, 'HUAWEI-WLAN-VAP-MIB', null, '-OQUs');
```

The second one avoids the merge_recursive, but does caching we don't need.

One way of the other, this table is not expected to be 1000 entries long. It is only the doubled SSID table, because of the 2 bands. And even in big networks, you don't have 1000 SSIDs on a single controller. I am not sure caching and/or merge will ever be an issue here.

Reply:
All index values must have an oid you can access from.  Look at the snmp table, it is a member of the table.  Does it return nothing from the captured data or the actual device?  Because it would not be in the captured data.

Reply:
Documentation confirms the behaviour. The `hwWlanSsid` is the index, and cannot be walked by itself. So no other choice than extracting (and converting) the index from num to str or the opposite.

OID | Object Name | Syntax | Max Access | Description | Implemented Specifications
-- | -- | -- | -- | -- | --
1.3.6.1.4.1.2011.6.139.17.1.2.1.1 | hwWlanSsid | OCTET STRING | **not-accessible** | This object indicates the table index. | This object is implemented as defined in the corresponding MIB file.
1.3.6.1.4.1.2011.6.139.17.1.2.1.2 | hwWlanSsid2gStaCnt | Integer32 | read-only | This object indicates the number of users at 2.4G frequency band. | This object is implemented as defined in the corresponding MIB file.

Reply:
Fun.

Reply:
Annoyingly fun :)